File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/dns/DNSSettingsViewModel.kt
package com.openu60.feature.router.dns

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.DNSConfig
import com.openu60.core.model.DNSParser
import com.openu60.core.model.DoHCacheEntry
import com.openu60.core.model.DoHParser
import com.openu60.core.model.DoHStatus
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

enum class DNSMode { AUTO, CUSTOM, DOH }

data class DNSSettingsState(
    val dnsConfig: DNSConfig = DNSConfig.empty,
    val dohStatus: DoHStatus = DoHStatus.empty,
    val cacheEntries: List<DoHCacheEntry> = emptyList(),
    val dnsMode: DNSMode = DNSMode.AUTO,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
)

@HiltViewModel
class DNSSettingsViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(DNSSettingsState())
    val state: StateFlow<DNSSettingsState> = _state.asStateFlow()

    fun updateDnsConfig(config: DNSConfig) {
        _state.value = _state.value.copy(dnsConfig = config)
    }

    fun setDnsMode(mode: DNSMode) {
        val previousMode = _state.value.dnsMode
        _state.value = _state.value.copy(dnsMode = mode)
        if (previousMode == DNSMode.DOH && mode != DNSMode.DOH && _state.value.dohStatus.enabled) {
            disableDoH()
        }
    }

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val dnsDeferred = async { agentClient.getJSON("/api/network/dns") }
                val dohDeferred = async {
                    try { agentClient.getJSON("/api/doh/status") } catch (_: Exception) { emptyMap() }
                }
                val dnsData = dnsDeferred.await()
                val dohData = dohDeferred.await()
                val dnsConfig = DNSParser.parse(dnsData)
                val dohStatus = if (dohData.isNotEmpty()) DoHParser.parse(dohData) else DoHStatus.empty
                val mode = when {
                    dohStatus.enabled -> DNSMode.DOH
                    dnsConfig.isManual -> DNSMode.CUSTOM
                    else -> DNSMode.AUTO
                }
                _state.value = _state.value.copy(
                    dnsConfig = dnsConfig,
                    dohStatus = dohStatus,
                    dnsMode = mode,
                    isLoading = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun saveDNS() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val c = _state.value.dnsConfig
                val dnsMode = when (_state.value.dnsMode) {
                    DNSMode.AUTO -> "auto"
                    DNSMode.CUSTOM -> "manual"
                    DNSMode.DOH -> "manual"
                }
                val params = mapOf(
                    "dns_mode" to dnsMode,
                    "prefer_dns_manual" to c.primaryDns,
                    "standby_dns_manual" to c.secondaryDns,
                    "ipv6_prefer_dns_manual" to c.ipv6PrimaryDns,
                    "ipv6_standby_dns_manual" to c.ipv6SecondaryDns,
                )
                agentClient.putJSON("/api/network/dns", params)
                _state.value = _state.value.copy(isLoading = false, message = "DNS settings saved", messageIsError = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) saveDNS() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun enableDoH() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.postJSON("/api/doh")
                _state.value = _state.value.copy(
                    dohStatus = _state.value.dohStatus.copy(enabled = true),
                    isLoading = false,
                    message = "DoH enabled",
                    messageIsError = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) enableDoH() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun disableDoH() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                agentClient.deleteJSON("/api/doh")
                _state.value = _state.value.copy(
                    dohStatus = _state.value.dohStatus.copy(enabled = false),
                    isLoading = false,
                    message = "DoH disabled",
                    messageIsError = false,
                )
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) disableDoH() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun fetchDohCache() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val list = agentClient.getJSONArray("/api/doh/cache")
                val entries = DoHParser.parseCacheEntries(list)
                _state.value = _state.value.copy(cacheEntries = entries, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) fetchDohCache() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}

File: mobile/ios/OpenU60/Features/RouterSettings/DNS/DNSSettingsViewModel.swift
import SwiftUI

enum DNSMode: Int, CaseIterable, Identifiable {
    case auto, custom, doh

    var id: Int { rawValue }

    var label: String {
        switch self {
        case .auto: "Auto"
        case .custom: "Custom"
        case .doh: "DoH"
        }
    }
}

@Observable
@MainActor
final class DNSSettingsViewModel {
    var config: DNSConfig = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false

    // Mode
    var selectedMode: DNSMode = .auto

    // Editable fields
    var editPrimary: String = ""
    var editSecondary: String = ""

    // DoH
    var doh: DoHStatus = .empty
    var editUpstream: String = ""
    var cacheEntries: [DoHCacheEntry] = []
    var showCacheInspector: Bool = false

    private let client: AgentClient
    private let authManager: AuthManager
    private var initialLoadDone = false

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        message = nil

        do {
            let data = try await client.getJSON("/api/router/dns")
            config = DNSParser.parse(data)
            if !initialLoadDone {
                editPrimary = config.primaryDns
                editSecondary = config.secondaryDns
            }
        } catch {
            showMessage("Failed to load DNS: \(error.localizedDescription)", isError: true)
        }

        // DoH status (independent, don't fail the whole refresh)
        do {
            let data = try await client.getJSON("/api/doh/status")
            doh = DoHParser.parse(data)
            if !initialLoadDone && editUpstream.isEmpty {
                editUpstream = doh.upstreamUrl
            }
            if doh.enabled {
                await refreshCache()
            }
        } catch {
            doh = .empty
            if message == nil {
                showMessage("DoH status unavailable", isError: false)
            }
        }

        // Derive mode from state only on initial load
        if !initialLoadDone {
            if doh.enabled {
                selectedMode = .doh
            } else if config.isManual {
                selectedMode = .custom
            } else {
                selectedMode = .auto
            }
            initialLoadDone = true
        }

        isLoading = false
    }

    func apply() async {
        switch selectedMode {
        case .auto:
            await applyAuto()
        case .custom:
            await applyCustom()
        case .doh:
            await applyDoH()
        }
    }

    private func applyAuto() async {
        isLoading = true
        do {
            let _ = try await client.putJSON("/api/router/dns", body: [
                "dns_mode": "auto",
                "prefer_dns_manual": "",
                "standby_dns_manual": ""
            ])
            // Always disable DoH (idempotent, prevents orphaned dnsmasq forwarding)
            let _ = try await client.postJSON("/api/doh/disable")
            showMessage("DNS set to Auto", isError: false)
            await refresh()
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
            isLoading = false
        }
    }

    private func isValidIPv4(_ str: String) -> Bool {
        let parts = str.split(separator: ".", omittingEmptySubsequences: false)
        guard parts.count == 4 else { return false }
        return parts.allSatisfy { part in
            guard let n = UInt16(part) else { return false }
            return n <= 255
        }
    }

    private func applyCustom() async {
        guard !editPrimary.isEmpty else {
            showMessage("Primary DNS cannot be empty", isError: true)
            return
        }
        guard isValidIPv4(editPrimary) else {
            showMessage("Invalid primary DNS address", isError: true)
            return
        }
        if !editSecondary.isEmpty {
            guard isValidIPv4(editSecondary) else {
                showMessage("Invalid secondary DNS address", isError: true)
                return
            }
        }
        isLoading = true
        do {
            let _ = try await client.putJSON("/api/router/dns", body: [
                "dns_mode": "manual",
                "prefer_dns_manual": editPrimary,
                "standby_dns_manual": editSecondary
            ])
            // Always disable DoH (idempotent, prevents orphaned dnsmasq forwarding)
            let _ = try await client.postJSON("/api/doh/disable")
            showMessage("DNS set to Custom (\(editPrimary))", isError: false)
            await refresh()
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
            isLoading = false
        }
    }

    private func applyDoH() async {
        guard !editPrimary.isEmpty else {
            showMessage("Primary DNS cannot be empty", isError: true)
            return
        }
        guard isValidIPv4(editPrimary) else {
            showMessage("Invalid primary DNS address", isError: true)
            return
        }
        if !editSecondary.isEmpty {
            guard isValidIPv4(editSecondary) else {
                showMessage("Invalid secondary DNS address", isError: true)
                return
            }
        }
        guard !editUpstream.isEmpty else {
            showMessage("DoH upstream URL cannot be empty", isError: true)
            return
        }
        isLoading = true
        do {
            // Set manual DNS (DoH local resolver)
            let _ = try await client.putJSON("/api/router/dns", body: [
                "dns_mode": "manual",
                "prefer_dns_manual": editPrimary,
                "standby_dns_manual": editSecondary
            ])
            // Update DoH upstream
            let _ = try await client.putJSON("/api/doh/config", body: [
                "upstream_url": editUpstream
            ])
            // Enable DoH if not running
            if !doh.enabled {
                let _ = try await client.postJSON("/api/doh/enable")
            }
            showMessage("DoH enabled with \(editUpstream)", isError: false)
            await refresh()
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
            isLoading = false
        }
    }

    func clearCache() async {
        isLoading = true
        do {
            let _ = try await client.postJSON("/api/doh/cache/clear")
            showMessage("DNS cache cleared", isError: false)
            await refresh()
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
            isLoading = false
        }
    }

    func refreshCache() async {
        do {
            let list = try await client.getJSONArray("/api/doh/cache")
            cacheEntries = DoHParser.parseCacheEntries(list)
        } catch {
            // cache fetch is best-effort
        }
    }

    func applyPreset(primary: String, secondary: String, upstream: String) {
        initialLoadDone = true
        if selectedMode == .auto {
            selectedMode = .custom
        }
        editPrimary = primary
        editSecondary = secondary
        editUpstream = upstream
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}
